### PR TITLE
feat: add portal to aa-signers

### PR DESCRIPTION
**Pull request by avasisht23**
# Pull Request Checklist

- [x] Did you add new tests and confirm existing tests pass? (`yarn test`)
- [x] Did you update relevant docs? (docs are found in the `site` folder, and guidelines for updating/adding docs can be found in the [contribution guide](https://github.com/alchemyplatform/aa-sdk/blob/main/CONTRIBUTING.md))
- [x] Do your commits follow the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) standard?
- [x] Does your PR title also follow the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) standard?
- [x] If you have a breaking change, is it [correctly reflected in your commit message](https://www.conventionalcommits.org/en/v1.0.0/#examples)? (e.g. `feat!: breaking change`)
- [x] Did you run lint (`yarn lint:check`) and fix any issues? (`yarn lint:write`)
- [x] Did you follow the [contribution guidelines](https://github.com/alchemyplatform/aa-sdk/blob/main/CONTRIBUTING.md)?

---

### Detailed summary

- Added `PortalSigner` to the `@alchemy/aa-signers` package.
- Updated `packages/signers/package.json` to include `@portal-hq/web` as a dependency.
- Added `PortalAuthenticationParams` and `PortalUserInfo` interfaces to `packages/signers/src/portal/types.ts`.
- Updated `templates/typescript/base.json` to include `"dom"` in the `lib` array and `"es2021"` as the `target` in the `compilerOptions`.
- Updated `site/.vitepress/config.ts` to include a new section for `Portal Signer` in the navigation menu.
- Added new Markdown files for `getAddress`, `signMessage`, `getAuthDetails`, and `authenticate` methods of `PortalSigner` in `site/packages/aa-signers/portal` directory.
- Updated `site/packages/aa-signers/portal/guides/portal.md` with installation instructions for `@portal-hq/web` SDK and updated code snippets.
- Added `PortalSigner` class to `packages/signers/src/portal/signer.ts` with methods for `getAddress`, `signMessage`, `signTypedData`, and `authenticate`.

> The following files were skipped due to too many changes: `packages/signers/src/portal/signer.ts`, `site/packages/aa-signers/portal/signTypedData.md`, `site/packages/aa-signers/portal/introduction.md`, `site/packages/aa-signers/portal/constructor.md`, `packages/signers/src/portal/__tests__/signer.test.ts`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

**Comment by avasisht23** [comment]
Current dependencies on/for this PR:
* main
  * **PR #300** <a href="https://app.graphite.dev/github/pr/alchemyplatform/aa-sdk/300?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #301** <a href="https://app.graphite.dev/github/pr/alchemyplatform/aa-sdk/301?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #303** <a href="https://app.graphite.dev/github/pr/alchemyplatform/aa-sdk/303?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/alchemyplatform/aa-sdk/303?utm_source=stack-comment).

**Comment by avasisht23** [comment]
@moldy530 @denniswon see this [post](https://alchemyinsights.slack.com/archives/C05DAP59URG/p1701714478958569) as to why this fails.

tldr; portal SDK doesn't export things properly. I've asked them to revise. When I go into node_modules myself and edit, build and tests pass.

File: packages/signers/src/portal/signer.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  };
+
+  signTypedData = (params: SignTypedDataParams) => {
+    if (!this.signer) throw new Error("Not authenticated");
+
+    return this.signer.signTypedData(params);
+  };
+
+  authenticate = async () => {
+    if (this.inner == null) throw new Error("No provider found");
```
Comment (moldy530) [inline packages/signers/src/portal/signer.ts:60]: is this even possible?

File: site/snippets/portal.ts [hunk 2]
```diff
@@ -1,21 +1,14 @@
...
-import { createWalletClient, custom } from "viem";
-import { polygonMumbai } from "viem/chains";
+import { PortalSigner } from "@alchemy/aa-signers";
+import { sepolia } from "viem/chains";
 
-const portalOptions = {
-  autoApprove: true,
-  gatewayConfig: `${polygonMumbai.rpcUrls.alchemy.http}/${process.env.ALCHEMY_API_KEY}`,
-  chainId: 80001,
-  host: process.env.PORTAL_WEB_HOST,
```
Comment (moldy530) [inline site/snippets/portal.ts:10]: is this not necessary anymore?

Comment (avasisht23) [reply site/snippets/portal.ts:10]: nope it was optional on their constructor params, but fair can add it back if necessary

File: templates/typescript/base.json [hunk 3]
```diff
@@ -2,7 +2,7 @@
   "$schema": "https://json.schemastore.org/tsconfig",
   "display": "Default",
   "compilerOptions": {
-    "lib": ["es2022"],
+    "lib": ["es2022", "dom"],
```
Comment (moldy530) [inline templates/typescript/base.json:5]: why do we need this?

File: packages/signers/src/portal/signer.ts (hunk 1 above)
Comment (avasisht23) [reply packages/signers/src/portal/signer.ts:60]: that's... a good q. guess not. i'll remove

File: templates/typescript/base.json (hunk 3 above)
Comment (avasisht23) [reply templates/typescript/base.json:5]: yeah I forgot to send the screenshot, but portal has a bunch of window/dom-related elements in their SDK. I had to add this to build, other was "window not defined" related errors when trying to build.

File: packages/signers/src/portal/types.ts [hunk 4]
```diff
@@ -0,0 +1,15 @@
+import type { Address } from "viem";
+
+export interface PortalAuthenticationParams {}
+
+// taken from Portal SDK since not exported
```
Comment (dscrobonia) [inline packages/signers/src/portal/types.ts:5]: We can export this for y'all!

Comment (dscrobonia) [reply packages/signers/src/portal/types.ts:5]: We'll queue up an update.

**Review by dscrobonia — COMMENTED** [review]
Looks good!

File: packages/signers/src/portal/signer.ts [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+  };
+
+  signTypedData = (params: SignTypedDataParams) => {
+    if (!this.signer) throw new Error("Not authenticated");
+
+    return this.signer.signTypedData(params);
+  };
+
+  authenticate = async () => {
+    this.signer = new WalletClientSigner(
```
Comment (moldy530) [inline packages/signers/src/portal/signer.ts:60]: why not just create this in the constructor and have authenticate return `this.getAuthDetails`?

**Review by moldy530 — APPROVED** [review]
had small nit, but LGTM otherwise

**Pull request merged**